pkg/awsutils/awsutils.go

```diff
@@ -918,6 +922,62 @@ func (cache *EC2InstanceMetadataCache) DeallocIPAddresses(eniID string, ips []st
 	return nil
 }
 
+func (cache *EC2InstanceMetadataCache) cleanUpLeakedENIs() {
+	log.Info("Attempting to clean up leaked ENIs")
+
+	networkInterfaces, err := cache.getFilteredListOfNetworkInterfaces()
+	if err != nil {
+		log.Errorf("awsutils: unable to obtain network interfaces.")
+		return
+	}
+
+	for _, networkInterface := range networkInterfaces {
+		// Verify the description starts with "aws-K8S-"
+		if !strings.HasPrefix(aws.StringValue(networkInterface.Description), eniDescriptionPrefix)  {
+			continue
+		}
+		// Delete the ENI.
+		eniID := aws.StringValue(networkInterface.NetworkInterfaceId)
+		err = cache.deleteENI(eniID, maxENIBackoffDelay)
+		if err != nil {
+			log.Warnf("Failed to clean up leaked ENI %s: %v", eniID, err)
+		}
+	}
+}
+
+func (cache *EC2InstanceMetadataCache) getFilteredListOfNetworkInterfaces() ([]*ec2.NetworkInterface, error) {
+	// The tag key has to be "node.k8s.amazonaws.com/instance_id"
+	tagFilter := &ec2.Filter{
+		Name: aws.String("tag-key"),
+		Values: []*string{
+			aws.String(eniNodeTagKey),
+		},
+	}
+	// Only fetch "available" ENIs.
+	statusFilter := &ec2.Filter{
+		Name: aws.String("status"),
+		Values: []*string{
+			aws.String("available"),
```

Comment:
How do we ensure ENIs that just got created and are still 'available' are excluded

Reply:
Isn't it better to just obtain ENIs not in-use?

Reply:
We tag the ENIs after they have been attached:

https://github.com/aws/amazon-vpc-cni-k8s/blob/e56efa2bbd91f4b906f70a678fc8a0e849d42270/pkg/awsutils/awsutils.go#L557-L565

Reply:
That said, maybe we should still tag it even if the attachment fails, since the delete just after can also fail... 🤔

Reply:
Going back to @vipulsabhaya 's initial comment, could you help me understand how this filtering logic ensures that we're only deleting ENIs that have leaked?

Reply:
We tag the ENI after we attach it to the instance. Filtering on the tag key helps us find the ENIs that were created and attached by the CNI.

That said, we should probably use the instance ID as well to verify that the instance is terminated, but that would also result in a lot more calls to the EC2 API.